### DISCUSSION TITLE
"refresh_token' or 'assertion' is not valid. The token was issued for a different client id"

### DISCUSSION THREAD

**Discussion by TNTUP** [ORelio/Minecraft-Console-Client#1866] [Q&A] [answered]
Hey there, I referred my friend to use the Minecraft Console Client and I think this issue happened when (his friend) logged in with the minecraft launcher it might have changed something which MCC expects a token but its not the right one? Can't tell how to reproduce it as I don't use the "microsoft" method but my friend is. Didn't make an issue because this is just a question and my friend solved it by deleting cache.ini. 

Hence why I don't use microsoft method, lol xD

Here's the error if needed:

Exception non gérée : System.Exception: AADSTS70000: The provided value for the input parameter 'refresh_token' or 'assertion' is not valid. The token was issued for a different client id
Trace ID:
Correlation ID:
Timestamp: 2021-12-23 03:41:47Z
   à MinecraftClient.Protocol.Microsoft.RequestToken(String postData)
   à MinecraftClient.Protocol.Microsoft.RefreshAccessToken(String refreshToken)
   à MinecraftClient.Protocol.ProtocolHandler.MicrosoftLoginRefresh(String refreshToken, SessionToken& session)
   à MinecraftClient.Program.InitializeClient()
   à MinecraftClient.Program.Main(String[] args)

**Comment by ORelio** [comment]
This is likely because MCC generates its own client ID. MCC can reuse the session token from Minecraft launcher for convenience, but if Microsoft starts checking the client ID, of course it will differ. Or the client ID changed because MCC re-generated its own session ID.

Those are suggestions but further investigations would be needed, and there is likely something we can fix on our side (for example, re-using client ID from the Minecraft launcher and/or storing client ID alongside session ID in `SessionCache.ini`).

**Comment by ReinforceZwei** [comment]
~~The client ID is not generated by MCC itself, but issued by MS Azure. MSA use it to identify applications that use their oauth API. Refresh token is tied to client ID, so MSA rejected the request when MCC tries to use that refresh token from Mojang with MCC's client ID. In other words, a refresh token from Mojang launcher cannot be used by MCC. (Unless we somehow got the client ID of Majang launcher, which is nearly impossible)~~

~~The solution would be not to fetch and use session from Majang launcher anymore.~~

Edit: See my answer below

**Reply by TNTUP** [reply]
So thats why when he used his alt on the launcher it did cause issues with MCC? Good thing I don't use the modern method for my alts it would be a pain with those errors XD

**Comment by ReinforceZwei** [comment, marked as answer]
After checking the code again, I found that the issue is not actually related to Mojang launcher but the `mcc` sign-in method. I guess your friend have switched sign-in method. Currently, the client ID used in `mcc` and `browser` sign-in method are different, but the refresh process always use MCC client ID. Hence the refresh token from `mcc` sign-in method cannot be used by MCC.

Saving the client ID in session cache file, as @ORelio suggested, should be a possible solution.

**Comment by TNTUP** [comment]
Not related to minecraft launcher, the PC where I ran the console clients throws the error aswell. (mcc method)

[code]
Cached session is invalid or expired.

Exception non gérée : System.Exception: AADSTS70000: The provided value for the input parameter 'refresh_token' or 'assertion' is not valid. The token was issued for a different client id
Trace ID: bc91620f-4314-4d54-aba7-6357f25c7b02
Correlation ID: 33c80e12-ae48-4200-b9f9-ea81fb9b4036
Timestamp: 2021-12-28 21:56:37Z
   à MinecraftClient.Protocol.Microsoft.RequestToken(String postData)
   à MinecraftClient.Protocol.Microsoft.RefreshAccessToken(String refreshToken)
   à MinecraftClient.Protocol.ProtocolHandler.MicrosoftLoginRefresh(String refreshToken, SessionToken& session)
   à MinecraftClient.Program.InitializeClient()
   à MinecraftClient.Program.Main(String[] args)
   [/code]
   
 Can't use it anymore, downgrading till 1.17.1 support

**Reply by TNTUP** [reply]
I have opened up an issue report as this happens to me without Minecraft Launcher in %appdata%/.minecraft
